File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
from collections import deque

T = int(input())
dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

def bfs(x, y):
    q = deque([(x, y)])
    visited[x][y] = 1

    while q:
        x, y = q.popleft()
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]
            if 0<=nx<N and 0<=ny<M and arr[nx][ny]==1 and visited[nx][ny]==0:
                q.append((nx, ny))
                visited[nx][ny] = 1

for _ in range(T):
    M, N, K = map(int, input().split())
    arr = [[0]*M for _ in range(N)]
    visited = [[0]*M for _ in range(N)]
    for _ in range(K):
        y, x = map(int, input().split())
        arr[x][y] = 1
    
    cnt = 0
    for x in range(N):
        for y in range(M):
            if arr[x][y]==1 and visited[x][y]==0:
                bfs(x, y)
                cnt += 1
    # print()
    # for i in range(N):
    #     print(visited[i])
    # print()
    
    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int M, N, K;

vector<vector<int>> arr;
vector<vector<int>> visited;

int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0};

void bfs(int x, int y){
    queue<pair<int, int>> q;
    q.push({x, y});
    visited[x][y] = 1;

    while(!q.empty()){
        auto[x, y] = q.front();
        q.pop();

        for(int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(0<=nx && nx<N && 0<=ny && ny<M && arr[nx][ny]==1 && visited[nx][ny]==0){
                q.push({nx, ny});
                visited[nx][ny] = 1;
            }
        }
    }

}

int main(void){
    int T;
    cin >> T;

    for(int t=0; t<T; t++){
        cin >> M >> N >> K;
        int cnt = 0;

        arr = vector<vector<int>>(N, vector<int>(M, 0));
        visited = vector<vector<int>>(N, vector<int>(M, 0));
        vector<pair<int, int>> loc;

        for(int i=0; i<K; i++){
            int x, y;
            cin >> y >> x;

            arr[x][y] = 1;
            loc.push_back({x, y});
        }

        for(int i=0; i<K; i++){
            auto[x, y] = loc[i];
            if(arr[x][y]==1 && visited[x][y]==0){
                cnt++;
                bfs(x, y);
            }
        }
        cout << cnt << endl;
    }
    return 0;
}